### PR TITLE
Embedder: embedding of multiple instances of a server object in a single process

### PR DESCRIPTION
**Pull request by sougou**
The Embedder is a new feature in servenv that supports instantiation of multiple instances of a server object within the same process.

A server object typically registers http URL handlers, status parts and stats variables. Such objects will generate panics if instantiated multiple times within the same process. The Embedder provides equivalent APIs. If used, it remaps the endpoints into different name-spaces.

In the case of URLs, and additional prefix is added: `/url` -> `/name/url`.
In the case of stats vars, an additional dimension gets added: `{"a": 1}` -> `{"name.a": 1}`

The TabletServer, which is multiply instantiated in vtcombo, has been changed to use the Embedder, and the related hacks have cleaned up. There is no additional overhead if a TabletServer is instantiated anonymously, like in the case of vttablet. In the embedded case, there is a negligible overhead in some stats cases because some variable access has to be wrapped with accessor functions.

With this change, vtcombo has a functional status page that points to individual status sub-urls for each tablet server it instantiates.

### CODE REVIEW

**Comment by sougou** [comment]
@deepthi to also review.

File: go/vt/servenv/status.go [hunk 1]
```diff
@@ -34,41 +34,46 @@ import (
...
-	Banner   string
-	Fragment string
-	F        func() interface{}
+// AddStatusPart adds a new section to status. frag is used as a
+// subtemplate of the template used to render /debug/status, and will
+// be executed using the value of invoking f at the time of the
+// /debug/status request. frag is parsed and executed with the
+// html/template package. Functions registered with AddStatusFuncs
+// may be used in the template.
+func AddStatusPart(banner, frag string, f func() interface{}) {
```
Comment (rafael) [inline go/vt/servenv/status.go]: nit - I would rename `frag`, to be  `fragment`. I spent a bit of time thinking how hand grenades relate to all these templates. :rofl:

File: go/vt/vtcombo/tablet_map.go [hunk 2]
```diff
@@ -293,7 +295,7 @@ func (itc *internalTabletConn) Execute(ctx context.Context, target *querypb.Targ
 	if err != nil {
 		return nil, tabletconn.ErrorFromGRPC(vterrors.ToGRPC(err))
 	}
-	return reply, nil
+	return reply.Copy(), nil
```
Comment (rafael) [inline go/vt/vtcombo/tablet_map.go]: This change plus the one in:
```
func (result *Result) Copy() *Result {
```
Does not seem related to this PR. Am I missing something here? If it's not related to this, can we move it to a different one ? 

It seems to me that commit `1f4d932e60e31b59dc4f859d48a11e7f2e9e438c` can be just a small PR.

File: go/vt/vttablet/heartbeat/reader.go [hunk 3]
```diff
@@ -68,7 +68,7 @@ type Reader struct {
 }
 
 // NewReader returns a new heartbeat reader.
-func NewReader(checker connpool.MySQLChecker, config tabletenv.TabletConfig) *Reader {
+func NewReader(tsv connpool.TabletService, config tabletenv.TabletConfig) *Reader {
```
Comment (rafael) [inline go/vt/vttablet/heartbeat/reader.go]: I think this rename is a bit confusing. When I saw `TabletService`, I was thinking in something that would implement an RPC interface. Later down in the PR, I understood what's going on, but I think we should try to find a better name.

File: go/vt/vttablet/tabletserver/replication_watcher.go [hunk 4]
```diff
@@ -53,31 +54,27 @@ type ReplicationWatcher struct {
 	eventToken *querypb.EventToken
 }
 
-var replOnce sync.Once
```
Comment (rafael) [inline go/vt/vttablet/tabletserver/replication_watcher.go]: Why we don't need this anymore?

File: go/vt/servenv/embedder.go [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+func findOrCreateEmbedder(instanceName string) *embedder {
+	if e, ok := embeds[instanceName]; ok {
+		return e
+	}
+	e := &embedder{
+		name:        instanceName,
+		prefix:      "",
+		handleFuncs: make(map[string]*handleFunc),
+	}
+	if instanceName != "" {
```
Comment (demmer) [inline go/vt/servenv/embedder.go]: I _think_ we should never get into here if instanceName is "", right?

That seems like a coding error and I'd panic instead of creating an empty embedder you never use.

File: go/vt/servenv/embedder.go [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+
+	if vmap, ok := globalStatVars[name]; ok {
+		vmap.Set(e.name, f)
+		return stats.NewCountersFuncWithMultiLabels("", help, labels, f)
+	}
+	vmap := &varMap{vars: map[string]func() map[string]int64{e.name: f}}
+	globalStatVars[name] = vmap
+
+	newlabels := append(append(make([]string, 0, len(labels)+1), e.prefix), labels...)
+	_ = stats.NewCountersFuncWithMultiLabels(e.prefix+name, help, newlabels, func() map[string]int64 {
```
Comment (demmer) [inline go/vt/servenv/embedder.go]: It seems wrong to me to include `e.prefix` in the name of the metric *and* in a label. Is this actually correct?

File: go/vt/servenv/embedder.go [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+	if instanceName == "" || name == "" {
+		return stats.NewCountersFuncWithMultiLabels(name, help, labels, f)
+	}
+
+	embedmu.Lock()
+	defer embedmu.Unlock()
+	e := findOrCreateEmbedder(instanceName)
+
+	if vmap, ok := globalStatVars[name]; ok {
+		vmap.Set(e.name, f)
```
Comment (demmer) [inline go/vt/servenv/embedder.go]: I'm confused about this... when two separate embedded instances register the same stat var (albeit with different instanceNames), won't this make it so the last one to register "wins" and overrides the other?

Logically, it seems to me we need to register ONE metric under `name` that then calls a function which collects the values from each instance (tagged with the instanceName) as one big CountersFuncWithMultiLabels.

**Comment by sougou** [comment]
After much discussion, we've decided that a better approach will be to reassemble parts of vttablet (mostly the conn pools) underneath vtgate.
This will be much easier once we've cleaned up the deprecated SBR code.

Closing this PR for now.
